### PR TITLE
Adding support for value = "width" or "height"  for x/x2/y/y2  channels.

### PR DESCRIPTION
**Pull request by kanitw**
We don't have a way to say `{"field": {"group": "width"/ "height"}}`  in Vega-Lite yet. 

I propose that we support `{x/x2/y/y2: {"value": "min" / "max"}}` for this. 

Another alternative keyword could be "start" / "end"?  Any thoughts?

Fix #https://github.com/vega/vega-lite/issues/3022

cc: @jheer @arvind

### CODE REVIEW

**Comment by kanitw** [comment]
image

```
{
  "$schema": "https://vega.github.io/schema/vega-lite/v2.json",
  "data": {"url": "data/seattle-weather.csv"},
  "layer": [
    {
      "mark": "bar",
      "encoding": {
        "x": {
          "timeUnit": "month",
          "field": "date",
          "type": "ordinal"

        },
        "y": {
          "aggregate": "mean",
          "field": "precipitation",
          "type": "quantitative"
        }
      }
    },
    {
      "mark": "rule",
      "encoding": {
        "y": {
          "value": "min"
        },
        "color": {"value": "red"},
        "size": {"value": 3}
      }
    }
  ]
}
```

**Comment by domoritz** [comment]
I think min and max can be confusing when you reverse the quantitative axis, right?

**Comment by kanitw** [comment]
For that case, wouldn't "start" and "end" also confusing?

**Comment by domoritz** [comment]
Probably but less so. Maybe `top`, `bottom`, `left`, `right` are better (but can't be transposed).

**Comment by kanitw** [comment]
How about `{"value": "width" / "height"}`?   (We actually only need one-side -- the other side `{value: 0}` is fine.)

**Comment by domoritz** [comment]
```json
{
  "$schema": "https://vega.github.io/schema/vega-lite/v2.json",
  "description": "A simple bar chart with embedded data.",
  "data": {
    "values": [
      {"a": "A","b": 28},
      {"a": "B","b": 55},
      {"a": "C","b": 43}
    ]
  },
  "encoding": {
    "y": {"field": "a", "type": "ordinal"},
    "x": {
      "field": "b",
      "type": "quantitative",
      "sort": "descending"
    }
  },
  "layer": [{
    "mark": "bar"
  }, {
    "mark": {
      "type": "text",
      "align": "right",
      "baseline": "middle",
      "dx": -4,
      "color": "white"
    },
    "encoding": {
      "text": {"field": "b", "type": "quantitative"},
      "x": {"value": "width"}
    }
  }]
}
```

image

**Comment by domoritz** [comment]
https://github.com/vega/vega-lite/issues/3775 could be a good example.

**Comment by kanitw** [comment]
layer_bar_annotation is the one that need this

**Comment by kanitw** [comment]
I think documenting this is a bit tricky since we have have specific section for value def for different type of marks.  So let's punt on that until we do https://github.com/vega/vega-lite/issues/3694, but at least now we have an example that uses this!

File: examples/specs/layer_bar_annotations.vl.json [hunk 1]
```diff
@@ -54,11 +54,14 @@
...
-          "align": "left",
-          "dx": 215,
-          "dy": -5
+          "align": "right",
+          "dx": -2,
+          "dy": -4
         },
         "encoding": {
+          "x": {
+            "value": "width"
```
Comment (domoritz) [inline examples/specs/layer_bar_annotations.vl.json]: :+1:

**Pull request merged**